text/0000-async_await.md

```diff
@@ -0,0 +1,571 @@
+statements). Given that, the only question is whether or not we could have a
+more expedited stability by using generators for the time being than by
+introducing async functions now.
+
+It seems unlikely that using macros which expand to generators will result in a
+faster stabilization. Generators can express a wider range of possibilities,
+and have a wider range of open questions - both syntactic and semantic. This
+does not even address the open questions of stabilizing more procedural macros.
+For this reason, we believe it is more expedient to stabilize the minimal
+built-in async/await functionality than to attempt to stabilize generators and
+proc macros.
+
+## `async` based on generators alone
+
+Another alternative design would be to have async functions *be* the syntax for
+creating generators. In this design, we would write a generator like this:
+
+```rust
+async fn foo(arg: Arg) -> Return yield Yield
+```
+
+Both return and yield would be optional, default to `()`. An async fn that
+yields `()` would implement `Future`, using a blanket impl. An async fn that
+returns `()` would implement `Iterator`.
+
+The problem with this approach is that does not ergonomically handle `Stream`s,
+which need to yield `Async<T>`. It's unclear how `await` inside of an async fn
+yielding something other than `()` (which would include streams) would work.
+For this reason, the "matrix" approach in which we have independent syntax for
+generator functions, async functions, and async generator functions, seems like
+a more promising approach.
+
+## "Hot async functions"
+
+As proposed by this RFC, all async functions return immediately, without
+evaluating their bodies at all. As discussed above, this is not convenient for
+use cases in which you have an immediate "initialization" step - those use
+cases need to use a terminal async block, for example.
+
+An alternative would be to have async functions immediately evaluate up until
+their first `await`, preserving their state until then. The implementation of
+this would be quite complicated - they would need to have an additional yield
+point within the `await`, prior to polling the future being awaited,
+conditional on whether or not the await is the first await in the body of the
+future.
+
+This is also complicated from a user perspective - when a portion of the body
+is evaluated depends on whether or not it appears before all `await`
+statements (which could possibly be macro generated). The use of a terminal
+async block provide a clearer mechanism for distinguishing between the
+immediately evaluated and asynchronously evaluated portions of a future with an
+initialization step.
+
+## Using async/await instead of alternative asynchronicity systems
+
+A final - and extreme - alternative would be to abandon futures and async/await
+as the mechanism for async/await in Rust and to adopt a different paradigm.
+Among those suggested are a generalized effects system, monads & do notation,
```

Comment:
First I want to say that I think this really well written (and readable) RFC. 🥇 
I think it also lands very nicely.

-----------

In particular, my initial thinking here, is this RFC is forward-compatible with forms of effect polymorphism; which is quite nice.

One could imagine `?async` as "possibly async", i.e: the callee decides on sync or async.
And `?const` as "possibly const", -- again, the callee decides.
Then we can mix these and you get 4 different modes.

Going a bit further (or too far) effects may be `try async` which could be the same as the Result-embedded futures encoding?

------------

Also, should `async fn(T) -> R` as a function type a thing?

Reply:
> Also, should async fn(T) -> R as a function type a thing?

Isn't that just `fn(T) -> impl Future<Item=R>`?

Reply:
That seemed to be the consensus on #rust-lang, yes =)

Reply:
Some further notes on the effect polymorphism angle (still noting that this is just "hypothetical"):

We could possibly interpret `async fn foo(x: T) -> R {..}` as the following:
`effect<E> (E + async) fn foo(x: T) -> R {..}`.
Which universally quantifies the ambient effect `E` and then you combine `E + async`.

But the body of `foo` can't assume anything about what `E` might be, so this should also be backwards compatible in the future with such as change?

Reply:
> The question was more: should the surface syntax be a thing?
with analogy with const fn(T) -> R.

Well, very much unlike `const`, `async` can be just syntactic sugar.